### ISSUE TITLE
ENOENT: no such file or directory, open '/usr/src/app/etc/public-suffix-list.tmp'

### ISSUE THREAD

**Issue by ekzyis** [haraka/haraka-tld#73]
Hey, I noticed that your package started to throw `ENOENT`. 

This started 15 days ago. 

I have found your code for the regular updates and it happened again today which makes sense since the interval is set to 15 days.

Error is thrown at line 37 in `lib/update.js`:

https://github.com/haraka/haraka-tld/blob/dc7b620d21bb3846cd58e3234b611b4c5a6836c7/lib/update.js#L36-L39

I am using the latest release, haraka-tld v1.0.34.

**Comment by ekzyis** [comment]
I was able to reproduce this by running `updatePSLfile()` using `ts-node`:

```
$ npx ts-node
> import { updatePSLfile } from "haraka-tld/lib/update";
undefined
> updatePSLfile();
Promise {
  <pending>,
  [Symbol(async_id_symbol)]: 42,
  [Symbol(trigger_async_id_symbol)]: 2,
  [Symbol(destroyed)]: { destroyed: false }
}
> Uncaught:
<ref *1> [Error: ENOENT: no such file or directory, open '/home/ekzyis/abusix/reporterservice/etc/public-suffix-list.tmp'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/home/ekzyis/abusix/reporterservice/etc/public-suffix-list.tmp',
  domainEmitter: WriteStream {
    fd: null,
    path: '/home/ekzyis/abusix/reporterservice/etc/public-suffix-list.tmp',
    flags: 'w',
    mode: 438,
    start: undefined,
    pos: undefined,
    bytesWritten: 0,
    closed: false,
    _writableState: WritableState {
      objectMode: false,
      highWaterMark: 16384,
      finalCalled: false,
      needDrain: false,
      ending: false,
      ended: false,
      finished: false,
      destroyed: true,
      decodeStrings: true,
      defaultEncoding: 'utf8',
      length: 0,
      writing: false,
      corked: 0,
      sync: true,
      bufferProcessing: false,
      onwrite: [Function: bound onwrite],
      writecb: null,
      writelen: 0,
      afterWriteTickInfo: null,
      buffered: [],
      bufferedIndex: 0,
      allBuffers: true,
      allNoop: true,
      pendingcb: 0,
      constructed: true,
      prefinished: false,
      errorEmitted: true,
      emitClose: true,
      autoDestroy: true,
      errored: [Circular *1],
      closed: true,
      closeEmitted: false,
      [Symbol(kOnFinished)]: []
    },
    _events: [Object: null prototype] {
      close: [Function (anonymous)],
      [Symbol(kConstruct)]: [Function]
    },
    _eventsCount: 2,
    _maxListeners: undefined,
    [Symbol(kFs)]: {
      appendFile: [Function: appendFile],
      appendFileSync: [Function: appendFileSync],
      access: [Function: access],
      accessSync: [Function: accessSync],
      chown: [Function: chown],
      chownSync: [Function: chownSync],
      chmod: [Function: chmod],
      chmodSync: [Function: chmodSync],
      close: [Function: close],
      closeSync: [Function: closeSync],
      copyFile: [Function: copyFile],
      copyFileSync: [Function: copyFileSync],
      cp: [Function: cp],
      cpSync: [Function: cpSync],
      createReadStream: [Function: createReadStream],
      createWriteStream: [Function: createWriteStream],
      exists: [Function: exists],
      existsSync: [Function: existsSync],
      fchown: [Function: fchown],
      fchownSync: [Function: fchownSync],
      fchmod: [Function: fchmod],
      fchmodSync: [Function: fchmodSync],
      fdatasync: [Function: fdatasync],
      fdatasyncSync: [Function: fdatasyncSync],
      fstat: [Function: fstat],
      fstatSync: [Function: fstatSync],
      fsync: [Function: fsync],
      fsyncSync: [Function: fsyncSync],
      ftruncate: [Function: ftruncate],
      ftruncateSync: [Function: ftruncateSync],
      futimes: [Function: futimes],
      futimesSync: [Function: futimesSync],
      lchown: [Function: lchown],
      lchownSync: [Function: lchownSync],
      lchmod: undefined,
      lchmodSync: undefined,
      link: [Function: link],
      linkSync: [Function: linkSync],
      lstat: [Function: lstat],
      lstatSync: [Function: lstatSync],
      lutimes: [Function: lutimes],
      lutimesSync: [Function: lutimesSync],
      mkdir: [Function: mkdir],
      mkdirSync: [Function: mkdirSync],
      mkdtemp: [Function: mkdtemp],
      mkdtempSync: [Function: mkdtempSync],
      open: [Function: open],
      openSync: [Function: openSync],
      opendir: [Function: opendir],
      opendirSync: [Function: opendirSync],
      readdir: [Function: readdir],
      readdirSync: [Function: readdirSync],
      read: [Function: read],
      readSync: [Function: readSync],
      readv: [Function: readv],
      readvSync: [Function: readvSync],
      readFile: [Function: readFile],
      readFileSync: [Function: readFileSync],
      readlink: [Function: readlink],
      readlinkSync: [Function: readlinkSync],
      realpath: [Function],
      realpathSync: [Function],
      rename: [Function: rename],
      renameSync: [Function: renameSync],
      rm: [Function: rm],
      rmSync: [Function: rmSync],
      rmdir: [Function: rmdir],
      rmdirSync: [Function: rmdirSync],
      stat: [Function: stat],
      statSync: [Function: statSync],
      symlink: [Function: symlink],
      symlinkSync: [Function: symlinkSync],
      truncate: [Function: truncate],
      truncateSync: [Function: truncateSync],
      unwatchFile: [Function: unwatchFile],
      unlink: [Function: unlink],
      unlinkSync: [Function: unlinkSync],
      utimes: [Function: utimes],
      utimesSync: [Function: utimesSync],
      watch: [Function: watch],
      watchFile: [Function: watchFile],
      writeFile: [Function: writeFile],
      writeFileSync: [Function: writeFileSync],
      write: [Function: write],
      writeSync: [Function: writeSync],
      writev: [Function: writev],
      writevSync: [Function: writevSync],
      Dir: [class Dir],
      Dirent: [class Dirent],
      Stats: [Function: Stats],
      ReadStream: [Getter/Setter],
      WriteStream: [Getter/Setter],
      FileReadStream: [Getter/Setter],
      FileWriteStream: [Getter/Setter],
      _toUnixTimestamp: [Function: toUnixTimestamp],
      F_OK: 0,
      R_OK: 4,
      W_OK: 2,
      X_OK: 1,
      constants: [Object: null prototype],
      promises: [Getter]
    },
    [Symbol(kIsPerformingIO)]: false,
    [Symbol(kCapture)]: false
  },
  domainThrown: false
}
> Uncaught:
[Error: ENOENT: no such file or directory, rename '/home/ekzyis/abusix/reporterservice/etc/public-suffix-list.tmp' -> '/home/ekzyis/abusix/reporterservice/etc/public-suffix-list'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'rename',
  path: '/home/ekzyis/abusix/reporterservice/etc/public-suffix-list.tmp',
  dest: '/home/ekzyis/abusix/reporterservice/etc/public-suffix-list'
}
```

**Comment by ekzyis** [comment]
The error goes away if I create the missing `etc` folder in the directory.

I think it is actually looking for the `etc` directory of this package. But somehow, it uses the project root folder instead of the package root folder.

**Comment by ekzyis** [comment]
#74 fixes this. 
I am not sure how this ever worked. But it definitely did in the past since I've been using this package for a while and only got this error now.
The automatic updater was added in c507a75 (2019). And `path.resolve` has _always_ been using the current working directory as far as I am aware. So I am really confused right now.

**Comment by msimerson** [comment]
> The error goes away if I create the missing etc folder in the project root directory.

Your test is calling the script differently than the built-in updater, which is called from `./index.js`. In your example, `path.resolve` is returning the current working directory with the requested path bits appended. In your manual example, the cwd is `/home/ekzyis/abusix/reporterservice`, which doesn't have an `etc` directory. When you create it, your CLI example works as you expect.

My guess is that something in your project is changing the cwd of the process, and that is breaking the updater.

**Comment by ekzyis** [comment]
> My guess is that something in your project is changing the cwd of the process, and that is breaking the updater.

I am using this code as a library, so my cwd is indeed not the root folder of the library where your `index.js` is but `/home/ekzyis/abusix/reporterservice`. So I'd expect my test / code example to run as it is. I shouldn't have to cd into the library folder. But maybe this code wasn't expected to be used as a library? That would explain it.

However, I still don't know why this broke now and not a year ago. I've been using this for like 2 years now, haha

But thanks for your fast response and merge!

**Comment by msimerson** [comment]
> But maybe this code wasn't expected to be used as a library?

It's not. Reaching into another projects / packages directory and running internal stuff *might* work, but it's definitely not a good practice and is likely to expose issues like this. When a package *intends* to export libraries, they usually do so on purpose, by documenting an API and exporting them in the packages main file (typically `./index.js').

**Comment by ekzyis** [comment]
Mhh, I think there is some misunderstanding going on here and want to clarify:

I am only using `get_organizational_domain` which _is_ exposed in `index.js`.
I only digged into the code and run this test code because I was getting an unexpected `ENOENT` after having used `get_organizational_domain` for two years now. Then I realized there is this automatic updater.
So I didn't mean to run internal stuff. I just wanted to debug the issue.

Hope this clears things up! :)
